File: A04 HashMap HashSet or related/related problems/leetcode13 RomanToInteger/RomantoInteger.py
class Solution:
    def romanToInt(self, s: str) -> int:
        roman_to_int = {
            'I' : 1,
            'V' : 5,
            'X' : 10,
            'L' : 50,
            'C' : 100,
            'D' : 500,
            'M' : 1000
        }
        
        result = 0
        prev_value = 0
        for letter in s:
            value = roman_to_int[letter]
            result += value
            if value > prev_value:
                # preceding roman nummber is smaller
                # we need to undo the previous addition
                # and substract the preceding roman char
                # from the current one, i.e. we need to
                # substract twice the previous roman char
                result -= 2 * prev_value
            prev_value = value
        return result


#################
class Solution:
    def romanToInt(self, s: str) -> int:
        translations = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        number = 0
        s = s.replace("IV", "IIII").replace("IX", "VIIII")
        s = s.replace("XL", "XXXX").replace("XC", "LXXXX")
        s = s.replace("CD", "CCCC").replace("CM", "DCCCC")
        for char in s:
            number += translations[char]
        return number

File: A04 HashMap HashSet or related/related problems/leetcode13 RomanToInteger/RomanToInteger.java
class Solution {
    public int romanToInt(String s) {
        int result = map(s.charAt(s.length() -1));   //private part; last character
        
        for (int i = s.length() -2 ; i >= 0 ; i--){
            if (map(s.charAt(i)) >= map(s.charAt(i+1))) {     // s.length() -2  > s.length()-1
                result += map(s.charAt(i));                
            } else{
                result -=map(s.charAt(i));
            }
        }
        
        return result;
    }
    
    
    private static int map(char c){
        switch(c){
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
            default: return 0;
        }
    }
}